### DISCUSSION TITLE
What is the (intended) way to do plugin based clean up, config._cleanup.append(func) vs pytest_unconfigure?

### DISCUSSION THREAD

**Discussion by symonk** [pytest-dev/pytest#7992] [General] [answered]
Hi, There seems to be two _similar_ mechanisms surrounding post-session and general cleanup.  Which is the preferred method for plugin related clean up,  outlined below is some sudo code:

```python
# Option 1
def pytest_configure(config) -> None:
    # do something if some condition is true, something I need undone
    # Optional append a cleanup function to the config._cleanup directly.
    # (this will be popped and invoked post-session later on).
```

```python
# Option 2
def pytest_configure(config) -> None:
    # do something if some condition is true, something I need undone

def pytest_unconfigure(config) -> None:
    # undo the thing myself here.
```

both are handled by the `config._ensure_configuration()` so does it matter? Which one should people be tending to use and is there any difference in both approaches?  Obviously to me the `unconfigure` hook is the way to go if you are a third party to pytest.

**Question:** When (working as a contributor) should we be loading things into the config._cleanup vs having the core plugin register its own unconfigure hook ? (third party users of pytest I assume should never be doing this as its private and possible to change?

**Comment by RonnyPfannschmidt** [comment, marked as answer]
we currently have no clear guide for that, and the api currently is not that nice

ideally configure/unconfigure turns into something that gets recorded in the "setupstate" but that refactoring has neither been discussed, nor explored yet

currently in core, using the cleanup is acceptable, plugins have to bite it and use configure/unconfigure
